crates/bevy_ecs/src/query/iter.rs

```diff
@@ -1716,6 +1725,9 @@ impl<'w, 's, D: QueryData, F: QueryFilter> QueryIterationCursor<'w, 's, D, F> {
                 if self.current_row == self.current_len {
                     let table_id = self.storage_id_iter.next()?.table_id;
                     let table = tables.get(table_id).debug_checked_unwrap();
+                    if table.is_empty() {
+                        continue;
+                    }
                     // SAFETY: `table` is from the world that `fetch/filter` were created for,
                     // `fetch_state`/`filter_state` are the states that `fetch/filter` were initialized with
                     unsafe {
```

Comment:
At the end of this `if` there is another `continue`. I think it was there to handle empty tables (since it would lead to repeating the above `if self.current_row == self.current_len` check), however now that they are skipped it should be useless.

Same applies to the `else` branch further down.

Reply:
but it did not skip the main loss caused by empty tables because `D::set_table/F::set_table` has already been executed before continuing next loop

Reply:
Yeah my point is not that it was there for efficiency, but for correctness (if it wasn't there the following code would try to obtain the first element of an empty archetype/table). My point is that now that `continue` is no longer needed because your check handles the same situation but better.